// javonet-nodejs-sdk/lib/sdk/ConfigRuntimeFactory.js
// @ts-check
import { WsConnectionData } from '../utils/connectionData/WsConnectionData.js'
import { TcpConnectionData } from '../utils/nodejs/connectionData/TcpConnectionData.js'
import { RuntimeName } from '../utils/RuntimeName.js'
import { RuntimeNameHandler } from '../utils/RuntimeNameHandler.js'
import { RuntimeContext } from './RuntimeContext.js'
import { isNodejsRuntime, isBrowserRuntime, getRequire, getDirname } from '../utils/Runtime.js'
import { InMemoryConnectionData } from '../utils/connectionData/InMemoryConnectionData.js'
import { JsonResolver } from './tools/JsonResolver.js'
import { UtilsConst } from '../utils/UtilsConst.js'

const requireDynamic = getRequire(import.meta.url)

/**
 * @typedef {import('../types.d.ts').ConfigSource} ConfigSource
 * @typedef {import('../types.d.ts').RuntimeName} RuntimeNameType
 */
/**
 * The ConfigRuntimeFactory class provides methods for creating runtime contexts.
 * Each method corresponds to a specific runtime (CLR, JVM, .NET Core, Perl, Ruby, Node.js, Python) and returns a RuntimeContext instance for that runtime.
 * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
 */
export class ConfigRuntimeFactory {
    /**
     * @param {ConfigSource | string} configSource
     */
    constructor(configSource) {
        this.configSource = configSource
    }

    static _IN_MEMORY_CONNECTION_TYPES = ['inmemory', 'memory']
    static _WEB_SOCKET_CONNECTION_TYPES = ['websocket', 'ws']
    static _TCP_CONNECTION_TYPES = ['tcp']

    /**
     * Creates RuntimeContext instance to interact with the .NET Framework runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the .NET Framework runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    clr(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Clr, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with the JVM runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the JVM runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    jvm(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Jvm, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with the .NET runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the .NET runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    netcore(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Netcore, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with the Perl runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the Perl runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    perl(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Perl, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with the Python runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the Python runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    python(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Python, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with the Ruby runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the Ruby runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    ruby(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Ruby, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with Node.js runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the Node.js runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    nodejs(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Nodejs, configName)
    }

    /** 
     * Creates RuntimeContext instance to interact with the Php runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the Php runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    php(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Php, configName)
    }

    /**
     * Creates RuntimeContext instance to interact with the Python 2.7 runtime.
     * @param {string} [configName="default"] - The name of the configuration to use (optional).
     * @return {RuntimeContext} a RuntimeContext instance for the Python 2.7 runtime
     * @see [Javonet Guides](https://www.javonet.com/guides/v2/javascript/foundations/runtime-context)
     */
    python27(configName = 'default') {
        return this.#getRuntimeContext(RuntimeName.Python27, configName)
    }

    /**
     * @param {RuntimeNameType} runtime
     * @param {string} configName
     * @returns {RuntimeContext}
     */
    #getRuntimeContext(runtime, configName = 'default') {
        const jsonResolver = new JsonResolver(this.configSource)

        try {
            UtilsConst.setLicenseKey(jsonResolver.getLicenseKey())
        } catch (e) {
            // licenseKey not found - ignore
        }

        try {
            UtilsConst.setJavonetWorkingDirectory(jsonResolver.getWorkingDirectory())
        } catch (e) {
            // workingDirectory not found - ignore
        }

        const channelType = jsonResolver.getChannelType(RuntimeNameHandler.getName(runtime), configName)
        const channelTypeLower = channelType ? channelType.trim().toLowerCase() : ''

        let connectionData = null

        // Browser runtime: accept only websocket types (mirror python behavior but safe for browser)
        if (isBrowserRuntime()) {
            if (ConfigRuntimeFactory._WEB_SOCKET_CONNECTION_TYPES.includes(channelTypeLower)) {
                connectionData = new WsConnectionData(
                    jsonResolver.getChannelHost(RuntimeNameHandler.getName(runtime), configName)
                )
            } else {
                throw new Error('Invalid connection type. Use inmemory, tcp or websocket')
            }

            const rtmCtx = RuntimeContext.getInstance(runtime, connectionData)
            this.#loadModules(runtime, configName, jsonResolver, rtmCtx)
            return rtmCtx
        }

        // Node runtime: support inmemory, tcp and websocket
        if (isNodejsRuntime()) {
            if (ConfigRuntimeFactory._IN_MEMORY_CONNECTION_TYPES.includes(channelTypeLower)) {
                connectionData = new InMemoryConnectionData()
            } else if (ConfigRuntimeFactory._TCP_CONNECTION_TYPES.includes(channelTypeLower)) {
                connectionData = new TcpConnectionData(
                    jsonResolver.getChannelHost(RuntimeNameHandler.getName(runtime), configName),
                    /** @type {number} */ (
                        jsonResolver.getChannelPort(RuntimeNameHandler.getName(runtime), configName)
                    )
                )
            } else if (ConfigRuntimeFactory._WEB_SOCKET_CONNECTION_TYPES.includes(channelTypeLower)) {
                connectionData = new WsConnectionData(
                    jsonResolver.getChannelHost(RuntimeNameHandler.getName(runtime), configName)
                )
            } else {
                throw new Error('Invalid connection type. Use inmemory, tcp or websocket')
            }

            const rtmCtx = RuntimeContext.getInstance(runtime, connectionData)
            this.#loadModules(runtime, configName, jsonResolver, rtmCtx)
            return rtmCtx
        }

        throw new Error('Unsupported runtime environment')
    }

    /**
     * @param {number} runtime
     * @param {string} configName
     * @param {JsonResolver} jsonResolver
     * @param {RuntimeContext} rtmCtx
     */
    #loadModules(runtime, configName, jsonResolver, rtmCtx) {
        try {
            const modules = jsonResolver
                .getModules(RuntimeNameHandler.getName(runtime), configName)
                .split(',')
                .map((m) => m.trim())
                .filter((m) => m !== '')

            if (!isNodejsRuntime()) {
                return
            }

            const path = /** @type {import('path').PlatformPath} */ (
                /** @type {unknown} */ (requireDynamic('path'))
            )

            let configDirectoryAbsolutePath = process.cwd()
            // support both possible property namings on JsonResolver
            const isCfgSourcePath = jsonResolver.isConfigSourcePath

            if (isCfgSourcePath) {
                // getDirname handles import.meta.url / file paths in different runtimes
                configDirectoryAbsolutePath = path.dirname(jsonResolver.configSource)
            }

            modules.forEach((module) => {
                if (path.isAbsolute(module)) {
                    rtmCtx.loadLibrary(module)
                } else {
                    rtmCtx.loadLibrary(path.join(configDirectoryAbsolutePath, module))
                }
            })
        } catch (error) {
            throw error
        }
    }
}
