// javonet-nodejs-sdk/lib/utils/connectionData/InMemoryConnectionData.js
// @ts-check
import { ConnectionType } from '../ConnectionType.js'
import { IConnectionData } from './IConnectionData.js'

/**
 * @typedef {import('../ConnectionType.js').ConnectionType} ConnectionType
 */

/**
 * @extends IConnectionData
 */
class InMemoryConnectionData extends IConnectionData {
    constructor() {
        super()
        this._connectionType = ConnectionType.IN_MEMORY
        this._hostname = 'inMemory'
    }

    /**
     * @returns {ConnectionType}
     */
    get connectionType() {
        return this._connectionType
    }

    /**
     * @returns {string}
     */
    get hostname() {
        return this._hostname
    }

    /**
     * Serializes the connection data.
     * @returns {number[]} An array of connection data values.
     */
    serializeConnectionData() {
        return [this.connectionType, 0, 0, 0, 0, 0, 0]
    }

    /**
     * @param {InMemoryConnectionData} other
     * @returns {boolean}
     */
    equals(other) {
        return other instanceof InMemoryConnectionData
    }

    /**
     * @returns {string}
     */
    toString() {
        return this.hostname
    }
}

export { InMemoryConnectionData }
