// javonet-nodejs-sdk/lib/sdk/tools/JsonResolver.js
//@ts-check
import { isNodejsRuntime, getRequire } from '../../utils/Runtime.js'

/**
 * @typedef {import('../../types.d.ts').RuntimeChannelType} RuntimeChannelType
 * @typedef {import('../../types.d.ts').ConfigSource} ConfigSource
 * @typedef {import('../../types.d.ts').RuntimeChannel} RuntimeChannel
 * @typedef {import('../../types.d.ts').Runtime} Runtime
 * @typedef {import('../../types.d.ts').Runtimes} Runtimes
 * @typedef {import('../../types.d.ts').RuntimeType} RuntimeType
 */

/**
 * @param {any} param
 * @returns {boolean}
 */
function isValidObject(param) {
    return typeof param === 'object' && param !== null && !Array.isArray(param)
}

/**
 * @param {string | ConfigSource} param
 * @returns {boolean}
 */
function isValidJson(param) {
    try {
        if (typeof param === 'string') {
            const parsed = JSON.parse(param)
            return typeof parsed === 'object' && parsed !== null
        }
        return false
    } catch (e) {
        return false
    }
}

const requireDynamic = getRequire(import.meta.url)

/** @type {any | null} */
let fs = null

class JsonResolver {
    /** @type {ConfigSource} */
    jsonObject
    /** @type {string} */
    configSource
    /** @type {boolean} */
    isConfigSourcePath = false

    /**
     * @param {string | ConfigSource} configSource
     */
    constructor(configSource) {
        this.configSource = JSON.stringify(configSource)
        this.jsonObject = this._getJsonObject(configSource)
    }

    /**
     * @returns {Partial<Runtimes>}
     */
    get runtimes() {
        return this.jsonObject?.runtimes || {}
    }

    /**
     * @returns {string}
     */
    get licenseKey() {
        return this.jsonObject?.licenseKey || 'your-license-key'
    }

    /**
     * @returns {string}
     */
    get workingDirectory() {
        return this.jsonObject?.workingDirectory || ''
    }

    /**
     * @returns {string}
     */
    getLicenseKey() {
        if (!this.licenseKey) {
            throw new Error('License key not found in configuration source. Check your configuration source.')
        }
        return this.licenseKey
    }

    /**
     * @returns {string}
     */
    getWorkingDirectory() {
        return this.workingDirectory
    }

    /**
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {RuntimeChannelType}
     */
    getChannelType(runtimeName, configName) {
        const channel = this._getChannel(runtimeName, configName)
        return channel.type
    }

    /**
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {string}
     */
    getChannelHost(runtimeName, configName) {
        const channel = this._getChannel(runtimeName, configName)
        if (channel.host === undefined) {
            throw new Error(`Host not found for channel in runtime ${runtimeName}, config ${configName}`)
        }
        return channel.host
    }

    /**
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {number|string}
     */
    getChannelPort(runtimeName, configName) {
        const channel = this._getChannel(runtimeName, configName)
        if (channel.port === undefined) {
            throw new Error(`Port not found for channel in runtime ${runtimeName}, config ${configName}`)
        }
        return channel.port
    }

    /**
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {string}
     */
    getModules(runtimeName, configName) {
        const runtime = this._getRuntime(runtimeName, configName)
        return runtime.modules || ''
    }

    /**
     * @private
     * @param {string | ConfigSource} configSource
     * @returns {ConfigSource}
     */
    _getJsonObject(configSource) {
        if (isValidObject(configSource)) {
            return /** @type {ConfigSource} */ (configSource)
        }
        if (typeof configSource === 'string' && isValidJson(configSource)) {
            const parsedJson = JSON.parse(configSource)
            if (isValidObject(parsedJson)) {
                return parsedJson
            }
        }
        if (isNodejsRuntime()) {
            if (!fs) {
                fs = requireDynamic('fs')
            }
            if (fs?.existsSync(configSource)) {
                this.isConfigSourcePath = true
                // read file content
                const jsonText = fs.readFileSync(configSource, 'utf8')
                if (isValidJson(jsonText)) {
                    const parsedFromFile = JSON.parse(jsonText)
                    if (isValidObject(parsedFromFile)) {
                        return parsedFromFile
                    }
                }
            }
        }

        throw new Error(
            `Configuration source is not valid. Check your configuration:${String(this.configSource)}`
        )
    }

    /**
     * @private
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {Runtime}
     */
    _getRuntime(runtimeName, configName) {
        const runtimeList = this.runtimes[runtimeName]

        if (Array.isArray(runtimeList)) {
            const runtime = runtimeList.find((item) => item.name === configName)
            if (runtime) {
                return runtime
            }
        } else if (typeof runtimeList?.name === 'string' && runtimeList?.name === configName) {
            return runtimeList
        }

        throw new Error(
            `Runtime config '${configName}' not found in configuration source for runtime '${runtimeName}'. Check your configuration source.`
        )
    }

    /**
     * @private
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {string}
     */
    _getRuntimeName(runtimeName, configName) {
        const runtime = this._getRuntime(runtimeName, configName)
        return runtime.name
    }

    /**
     * @private
     * @param {RuntimeType} runtimeName
     * @param {string} configName
     * @returns {RuntimeChannel}
     */
    _getChannel(runtimeName, configName) {
        const runtime = this._getRuntime(runtimeName, configName)
        if (!runtime.channel) {
            throw new Error(
                `Channel not found for runtime config '${configName}' and runtime '${runtimeName}'. Check your configuration source.`
            )
        }
        return runtime.channel
    }
}

export { JsonResolver }
