// javonet-nodejs-sdk/lib/utils/Runtime.js
// javascript
/** @typedef {import('../types.js').ConfigSource} ConfigSource */

import { createRequire } from './CreateRequire.node.js'

/** @type {import('url').pathToFileURL | null} */
let pathToFileURL = null
/** @type {import('url').fileURLToPath | null} */
let fileURLToPath = null
/** @type {import('path').PlatformPath | null} */
let path = null

/**
 * Validate if the runtime is Node.js
 * @returns {boolean}
 */
export function isNodejsRuntime() {
    return typeof process !== 'undefined' && process.versions != null && process.versions.node != null
}

/**
 * Validate if the runtime is Browser
 * @returns {boolean}
 */
export function isBrowserRuntime() {
    return typeof window !== 'undefined' && typeof window.document !== 'undefined'
}

/**
 * Validate if the runtime is ESM
 * @returns {boolean}
 */
export function isEsmRuntime() {
    return typeof __filename === 'undefined'
}

/**
 * Get the runtime extension
 * @returns {string}
 */
export function getRuntimeExtension() {
    return isEsmRuntime() ? 'js' : 'cjs'
}

/**
 * Get the require function
 * @param {string} [callerPath] - The path to the caller
 * @returns {any}
 */
export const getRequire = (callerPath) => {
    try {
        if (isNodejsRuntime()) {
            if (typeof require !== 'undefined') {
                return require
            }
            if (callerPath === undefined) {
                return createRequire(import.meta.url)
            }
            const baseUrl = callerPath?.startsWith('file:') ? callerPath : pathToFileURL(callerPath).href
            return createRequire(baseUrl)
        }
        return new Error('Runtime not supported')
    } catch (error) {
        throw error
    }
}

/**
 * Get current directory name in both CJS and ESM
 * @returns {string | Error}
 */
export const getDirname = () => {
    if (isBrowserRuntime()) {
        return new Error('Runtime not supported')
    }
    // when esm runtime, __dirname is undefined
    if (isEsmRuntime()) {
        const __filename = fileURLToPath(import.meta.url)
        return path.dirname(__filename)
    }
    // eslint-disable-next-line no-undef
    return __dirname
}

/**
 * Resolves the path to a dependency using Node.js module resolution.
 * @param {string} dependencyName - The name or path of the dependency to resolve.
 * @param {string} [callerPathOrUrl] - The `import.meta.url` (ESM) or `__filename` (CJS) of the calling module.
 * @returns {string} The resolved path to the dependency.
 * @throws {Error} If the dependency cannot be resolved.
 */
export const getDependency = (dependencyName, callerPathOrUrl) => {
    try {
        let baseUrl = callerPathOrUrl
        if (!baseUrl && typeof __filename !== 'undefined') {
            // eslint-disable-next-line no-undef
            baseUrl = __filename
        }

        const contextualRequire = getRequire(baseUrl)
        return contextualRequire.resolve(dependencyName)
    } catch (error) {
        throw new Error(
            `Failed to resolve dependency '${dependencyName}' from '${callerPathOrUrl}': ${error.message}`
        )
    }
}

if (isNodejsRuntime()) {
    const requireDynamic = getRequire(import.meta.url)
    if (fileURLToPath === null) {
        fileURLToPath = requireDynamic('url').fileURLToPath
    }
    if (path === null) {
        path = requireDynamic('path')
    }
    if (pathToFileURL === null) {
        pathToFileURL = requireDynamic('url').pathToFileURL
    }
}
