// javonet-nodejs-sdk/lib/core/namespaceCache/NamespaceCache.js
// @ts-check

class NamespaceCache {
    /** @type {NamespaceCache | null} */
    static _instance = null
    /** @type {string[]} */
    namespaceCache = []

    constructor() {
        if (NamespaceCache._instance === null) {
            NamespaceCache._instance = this
        }
        return NamespaceCache._instance
    }

    /**
     * @param {string} namespaceRegex
     */
    cacheNamespace(namespaceRegex) {
        this.namespaceCache.push(namespaceRegex)
    }

    /**
     * @returns {boolean}
     */
    isNamespaceCacheEmpty() {
        return this.namespaceCache.length === 0
    }

    /**
     * @param {string} typeToCheck
     * @returns {boolean}
     */
    isTypeAllowed(typeToCheck) {
        for (let pattern of this.namespaceCache) {
            if (new RegExp(pattern).test(typeToCheck.constructor.name)) {
                return true
            }
        }
        return false
    }

    getCachedNamespaces() {
        return this.namespaceCache
    }

    clearCache() {
        this.namespaceCache = []
        return 0
    }
}

export { NamespaceCache }
