// javonet-nodejs-sdk/lib/core/transmitter/TransmitterWrapper.js
// @ts-check
import { getDependency, getRequire, isNodejsRuntime } from '../../utils/Runtime.js'
import { ReceiverNative } from '../receiver/ReceiverNative.js'

/** @type {any | null} */
let library = null
/** @type {any | null} */
let path = null

/** @type {import('module').createRequire} */
const requireDynamic = getRequire(import.meta.url)

class TransmitterWrapper {
    static isNativeLibraryLoaded() {
        return Boolean(library)
    }

    static loadNativeLibrary() {
        if (!isNodejsRuntime()) {
            throw new Error('Javonet.loadNativeLibrary is not supported in this runtime')
        }
        if (this.isNativeLibraryLoaded()) {
            return
        }
        try {
            if (!path) {
                path = requireDynamic('path')
            }
            const packagePath = getDependency('javonet-binaries', import.meta.url)

            const binariesRootPath = path.dirname(packagePath)
            const nativeAddonPath = path.join(
                binariesRootPath,
                'build',
                'Release',
                'JavonetNodejsRuntimeAddon.node'
            )

            library = requireDynamic(nativeAddonPath)
            library.initializeTransmitter(binariesRootPath)
            // @ts-expect-error
            library.setReceiverNative(global?.ReceiverNative)
        } catch (error) {
            throw error
        }
    }

    /**
     * @param {string} licenseKey
     */
    static activate(licenseKey) {
        this.loadNativeLibrary()
        return library.activate(licenseKey)
    }

    /**
     * @param {Uint8Array} messageArray
     * @returns {Uint8Array}
     */
    static sendCommand(messageArray) {
        this.loadNativeLibrary()
        return library.sendCommand(messageArray)
    }

    /**
     * @param {string} configSource
     */
    static setConfigSource(configSource) {
        this.loadNativeLibrary()
        return library.setConfigSource(configSource)
    }

    /**
     * @param {string} path
     */
    static setJavonetWorkingDirectory(path) {
        this.loadNativeLibrary()
        library.setJavonetWorkingDirectory(path)
    }
}

export { TransmitterWrapper }
